Plutarch/DataRepr/Internal.hs

```diff
@@ -344,8 +385,77 @@ instance
       mkSOP :: NP (Term s) r -> SOP (Term s) (PCode s a)
       mkSOP = SOP . mkSum @_ @n @pcode
 
-pasDataSum :: PIsDataRepr a => Term s a -> Term s (PDataSum (PIsDataReprRepr a))
-pasDataSum = punsafeCoerce
+data LTReprHandlers (defs :: [[PLabeledType]]) (s :: S) where
+  LTRHNil :: LTReprHandlers '[] s
+  LTRHCons ::
+    (Term s (PDataRecord def) -> Term s (PDataRecord def) -> Term s PBool) ->
+    LTReprHandlers defs s ->
+    LTReprHandlers (def : defs) s
+
+-- | Optimized dual pmatch specialized for lexicographic '#<' (and potentially) '#<=' implementation.
+pmatchLT :: Term s (PDataSum defs) -> Term s (PDataSum defs) -> LTReprHandlers defs s -> Term s PBool
+pmatchLT d1 d2 (LTRHCons handler LTRHNil) = handler (punDataSum # d1) (punDataSum # d2)
+pmatchLT d1 d2 handlers = unTermCont $ do
+  a <- tcont . plet $ pasConstr #$ pforgetData $ pdata d1
+  b <- tcont . plet $ pasConstr #$ pforgetData $ pdata d2
+
+  cid1 <- tcont . plet $ pfstBuiltin # a
+  cid2 <- tcont . plet $ pfstBuiltin # b
+
+  pure $
+    pif
+      (cid1 #< cid2)
+      -- Left arg's constructor id is less, no need to continue.
+      (pconstant True)
+      $ pif
+        (cid1 #== cid2)
+        -- Matching constructors, compare fields now.
+        ( plet (psndBuiltin # a) $ \flds1 ->
+            plet (psndBuiltin # b) $ \flds2 ->
+              go 0 (applyHandlers flds1 flds2 handlers) cid1
+        )
+        -- Left arg's constructor id is greater, no need to continue.
+        $ pconstant False
+  where
+    applyHandlers :: Term s (PBuiltinList PData) -> Term s (PBuiltinList PData) -> LTReprHandlers defs s -> [Term s PBool]
+    applyHandlers _ _ LTRHNil = []
+    applyHandlers args1 args2 (LTRHCons handler rest) =
+      handler (punsafeCoerce args1) (punsafeCoerce args2) :
+      applyHandlers args1 args2 rest
+
+    go ::
```

Comment:
You don't optimise out any shared cases. This can happen if a sum type has two identical constructors, e.g. `PMaybeData PData PData`. Not sure this is worth optimising for.

Reply:
You can see how it's done in `pmatchDataSum`.

Reply:
You're completely duplicating parts of the logic in the same file.